### PR TITLE
REF: do less in Grouping.__init__

### PR DESCRIPTION
**Pull request by jbrockmendel**
making it incrementally easier to reason about what `self.grouper` is

### CODE REVIEW

File: pandas/core/groupby/grouper.py [hunk 1]
```diff
@@ -509,37 +510,24 @@ def __init__(
...
+                    # i.e. DataFrame case reachable if columns non-unique
+                    t = self.name or str(type(self.grouper))
+                    raise ValueError(f"Grouper for '{t}' not 1-dimensional")
+
             elif isinstance(self.grouper, (list, tuple)):
                 self.grouper = com.asarray_tuplesafe(self.grouper)
 
             # a passed Categorical
             elif is_categorical_dtype(self.grouper):
+                self._passed_categorical = True
```
Comment (jreback) [inline pandas/core/groupby/grouper.py:523]: why don't you just make this a property of the grouper itself?

File: pandas/core/groupby/grouper.py [hunk 2]
```diff
@@ -577,8 +565,20 @@ def __repr__(self) -> str:
     def __iter__(self):
         return iter(self.indices)
 
-    _codes: np.ndarray | None = None
-    _group_index: Index | None = None
+    @cache_readonly
+    def _ilevel(self) -> int | None:
```
Comment (jreback) [inline pandas/core/groupby/grouper.py:569]: why is this needed?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/grouper.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py:569]: i find it nicer as a property than re-calculating it in places where it is used

File: pandas/core/groupby/grouper.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py:523]: i think what you're suggesting is something like

```
@property
def _passed_categorical(self):
    return is_categorical_dtype(self.grouper)
```

that is the longer-term goal, but ATM we treat directly-passed Categoricals differently from Categoricals derived within `__init__`, so that property wouldn't be _quite_ the same as this variable.  I expect that the different treatment is non-intentional, but need to track that down to be sure.

**Comment by jbrockmendel** [comment]
rebased + green

**Comment by jbrockmendel** [comment]
@jreback gentle ping; this has become a blocker for #41373

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks. not super happy with adding state, but as you are working towards more robust ways this is fine.
